### PR TITLE
Feature pipeline integration tests

### PR DESCRIPTION
**Pull request by gusmith**
That was quite painful, but it is finally working end to end!!!
The integration step using docker compose has been removed to run these tests only using kubernetes.
We were otherwise running twice the same tests on different deployment.
With this pipeline, we are still testing the docker compose and the kubernetes deployment, but running different tests on each.
On the docker compose deployment, we run the benchmark and the tutorial tests.
On k8s, we run the integration tests. 
It was preferred as the integration tests are BIG, and takes a long time, so it is better t use some beefy infrastructure instead of a single small VM running the whole service (which was tricky because of timeouts).

Note: this is **not** deploying a release, which is out of scope for this work.
Note 2: this is to update the PR #388 removing the integration tests and running them in k8s only.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -160,55 +141,78 @@ stages:
...
+    - script: |
+        echo "Create Persistent Volume Claim"
+        cat .azurePipeline/k8s_test_pvc.yaml.tmpl | \
+        sed 's|\$PVC'"|$(PVC)|g" | \
+        sed 's|\$DEPLOYMENT_NAME'"|$(DEPLOYMENT)|g" | \
+        kubectl apply --wait=true -n=$(NAMESPACE) -f -
+      env:
+        KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Create persistent volume claim for future results.'
+    # Starting the whole entity-service
```
Comment (hardbyte) [inline azure-pipelines.yml]: Idea for an additional step before deploying to a test cluster: lint/test the helm chart. E.g. https://helm.sh/blog/chart-testing-intro/#example-using-it-with-circleci

File: azure-pipelines.yml [hunk 2]
```diff
@@ -160,55 +141,78 @@ stages:
...
+    # Task to get the IP of the service, which is require to run the test.
+    # Note that the port 80 is added at the end. This is in fact required for the command `dockerize -wait` to wait for a service before starting one.
     - script: |
-        kubectl get services -n=$(NAMESPACE) -lapp=$(DEPLOYMENT)-entity-service -o jsonpath="{.items[0].spec.clusterIP}" | xargs -I@ echo "##vso[task.setvariable variable=SERVICE_IP]@"
+        kubectl get services -n=$(NAMESPACE) -lapp=$(DEPLOYMENT)-entity-service -o jsonpath="{.items[0].spec.clusterIP}" | xargs -I@ echo "##vso[task.setvariable variable=SERVICE_IP]@:80"
       env:
         KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Set a variable for the service IP'
+    # The file `.azurePipeline/k8s_test_job.yaml.tmpl` is a template for the creation of the test job.
+    # The test job stops when all the tests have been ran, creating a test result file in the PVC.
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
    # The test job stops when all the tests have been run, creating a test result file in the PVC.
```

File: azure-pipelines.yml [hunk 3]
```diff
@@ -160,55 +141,78 @@ stages:
...
+        --set api.app.debug=true \
+        --set global.postgresql.postgresqlPassword=notaproductionpassword \
+        --set api.ingress.enabled=false \
+        --set api.www.image.tag=$(DOCKER_TAG) \
+        --set api.app.image.tag=$(DOCKER_TAG) \
+        --set api.dbinit.image.tag=$(DOCKER_TAG) \
+        --set workers.image.tag=$(DOCKER_TAG)
       env:
         KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Install the deployment on kubernetes'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Kubernetes deployment'
```

File: azure-pipelines.yml [hunk 4]
```diff
@@ -160,55 +141,78 @@ stages:
...
+        kubectl apply --wait=true -n=$(NAMESPACE) -f -
+        kubectl get pods -n=$(NAMESPACE) -l deployment=$(DEPLOYMENT) -o jsonpath="{.items[0].metadata.name}" | xargs -I@ kubectl wait -n=$(NAMESPACE) --timeout=300s --for=condition=Ready pods/@
         kubectl get pods -n=$(NAMESPACE) -l deployment=$(DEPLOYMENT) -o jsonpath="{.items[0].metadata.name}" | xargs -I@ kubectl logs -n=$(NAMESPACE) -f @
       env:
         KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Start the test job'
+    # That is still something fun with kubernetes, we cannot copy a file from a stopped container in a pod (while this is possible with docker), which is
+    # why we are creating a persistent volume, running the test job using the volume, and finally starting a container to be up with the mounted
+    # volume and use `kubectl cp` on it while it still runs.
+    # This is tracked in https://github.com/kubernetes/kubectl/issues/454 but has been in April 2018, its priority raised in June 2019.
```
Comment (hardbyte) [inline azure-pipelines.yml]: We could save the results to minio/s3 instead (which has to be configured or provisioned for the service to work).

File: azure-pipelines.yml [hunk 5]
```diff
@@ -221,13 +225,16 @@ stages:
...
+      displayName: 'Copy the results from the test job pod.'
+    # Finally, we can publish the test results.
     - task: PublishTestResults@2
       condition: succeeded()
       inputs:
         testResultsFormat: 'JUnit'
         testResultsFiles: 'k8s-results.xml'
         testRunTitle: 'Publish kubernetes test results'
         failTaskOnFailedTests: true 
+    # And do a lot of cleaning.
```
Comment (hardbyte) [inline azure-pipelines.yml]: All the cleaning could be done in one step if everything shared a label?

File: docs/devops.rst [hunk 6]
```diff
@@ -0,0 +1,35 @@
+Devops
+===========
+
+Azure Pipeline
+--------------
+
+``entity-service`` is automatically built and tested using Azure Pipeline
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
``anonlink-entity-service`` is automatically built and tested using Azure DevOps
```

File: docs/devops.rst [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+in the project `Anonlink <https://dev.azure.com/data61/Anonlink>`.
+
+It consists only of a `build pipeline <https://dev.azure.com/data61/Anonlink/_build?definitionId=1>`.
+
+The build pipeline is described by the script `azurePipeline.yml`
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
The build pipeline is defined by the script `azurePipeline.yml`
```

File: docs/devops.rst [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+which is using resources from the folder `.azurePipeline`.
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
which uses resources from the folder `.azurePipeline`.
```

File: docs/devops.rst [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+Mainly, it:
+
+- builds and pushes docker images
+  - the frontend ``data61/anonlink-nginx``
+  - the backend ``data61/anonlink-app``
+  - the tutorials ``data61/anonlink-docs-tutorials`` (used to tests the tutorial Python Notebooks)
+  - the benchmark ``data61/anonlink-benchmark`` (used to run the benchmark)
+- runs the benchmark using ``docker-compose`` and publishes the artiafct in Azure
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
- runs the benchmark using ``docker-compose`` and publishes the results as an artifact in Azure
```

File: docs/devops.rst [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+The build pipeline is triggered for every push on every branch. It is not triggered by Pull
+Requests not to build twice the same commit.
+
+The build pipeline requires two environment variables provided by Azure environment:
+
+- `dockerHubId`: username for the pipeline to push images to Data61 dockerhub
+- `dockerHubPassword`: password for the corresponding username (this is a secret variable).
+
+It also requires a secret file (given in the Azure `Library` tab) containing the ``aws`` cluster
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
It also requires a secret file (given in the Azure `Library` tab) containing the ``k8s`` cluster
```

Although we happen to be using AWS I don't think this is aws specific.

File: docs/devops.rst [hunk 11]
```diff
@@ -0,0 +1,35 @@
+Devops
+===========
+
+Azure Pipeline
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
Continuous Integration
```

File: azure-pipelines.yml [hunk 12]
```diff
@@ -133,6 +111,7 @@ stages:
     - template: .azurePipeline/templateDockerTagFromBranch.yml
     - script: |
         ./.azurePipeline/runDockerComposeTests.sh --no-ansi -p es$(DOCKER_TAG)$(Build.SourceVersion) -t $(DOCKER_TAG) -o $(resultFile) --type tutorials
+      displayName: 'Start docker compose tutorials tests'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Test notebook tutorials'
```

File: azure-pipelines.yml [hunk 13]
```diff
@@ -160,55 +141,78 @@ stages:
...
     - template: .azurePipeline/templateDockerTagFromBranch.yml
+    # We are selecting a deployment name to be unique, not to have collision when pushing multiple times on the same branch
+    # So we are using the git commit hash. BUT, a k8s deployment MUST start with a letter, so added `es` to start its name, 
+    # then it needs to be short enough not to fail other tags built from it. In fact, most of the created objects from the templates
+    # will use the deployment name as a prefix for their tags, adding extra suffixes. If the deployment name is too long, 
+    # the tags may be longer than 63 characters which can fail the deployment with some warnings hard to see.
     - script: |
-        echo "##vso[task.setvariable variable=DEPLOYMENT]deployment-$(DOCKER_TAG)$(Build.SourceVersion)"
+        echo es$(Build.SourceVersion)| head -c 10 | xargs -I@ echo "##vso[task.setvariable variable=DEPLOYMENT]@"
+      displayName: 'Set some variables part 1'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Set variables part 1'
```

File: azure-pipelines.yml [hunk 14]
```diff
@@ -160,55 +141,78 @@ stages:
...
+    # The task setting variables is split in two, because we cannot re-use directly a set variable in the same task
+    # (in this case the value of $(DEPLOYMENT)).
     - script: |
         echo "##vso[task.setvariable variable=KUBECONFIGFILE]$(DownloadSecureFile.secureFilePath)"
         echo "##vso[task.setvariable variable=PVC]$(DEPLOYMENT)-test-results"
         echo $(backendImageName):$(DOCKER_TAG) | xargs -I@ echo "##vso[task.setvariable variable=IMAGE_NAME_WITH_TAG]@"
-        echo $(DOCKER_TAG)-tmppod | xargs -I@ echo "##vso[task.setvariable variable=POD_NAME]@"
-    # Start the whole service using helm. We are currently lucky, the helm version on azure is compatible with the one on our cluster, but the latest version of helm (Client 2.14.1) is not (tested from my laptop).
+        echo $(DEPLOYMENT)-tmppod | xargs -I@ echo "##vso[task.setvariable variable=POD_NAME]@"
+      displayName: 'Set some variables part 2'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Set variables part 2'
```

File: azure-pipelines.yml [hunk 15]
```diff
@@ -160,55 +141,78 @@ stages:
...
+    # The file `.azurePipeline/k8s_test_pvc.yaml.tmpl` is a template for the creation of persistent volume via kubectl
+    - script: |
+        echo "Create Persistent Volume Claim"
+        cat .azurePipeline/k8s_test_pvc.yaml.tmpl | \
+        sed 's|\$PVC'"|$(PVC)|g" | \
+        sed 's|\$DEPLOYMENT_NAME'"|$(DEPLOYMENT)|g" | \
+        kubectl apply --wait=true -n=$(NAMESPACE) -f -
+      env:
+        KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Create persistent volume claim for future results.'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Create persistent volume claim for integration test results.'
```

File: azure-pipelines.yml [hunk 16]
```diff
@@ -160,55 +141,78 @@ stages:
...
         sed 's|\$IMAGE_NAME_WITH_TAG'"|$(IMAGE_NAME_WITH_TAG)|g" | \
         sed 's|\$SERVICE_IP'"|$(SERVICE_IP)|g" | \
-        kubectl apply -n=$(NAMESPACE) -f -
-        kubectl get pods -n=$(NAMESPACE) -l deployment=$(DEPLOYMENT) -o jsonpath="{.items[0].metadata.name}" | xargs -I@ kubectl wait -n=$(NAMESPACE) --timeout=120s --for=condition=Ready pods/@
+        kubectl apply --wait=true -n=$(NAMESPACE) -f -
+        kubectl get pods -n=$(NAMESPACE) -l deployment=$(DEPLOYMENT) -o jsonpath="{.items[0].metadata.name}" | xargs -I@ kubectl wait -n=$(NAMESPACE) --timeout=300s --for=condition=Ready pods/@
         kubectl get pods -n=$(NAMESPACE) -l deployment=$(DEPLOYMENT) -o jsonpath="{.items[0].metadata.name}" | xargs -I@ kubectl logs -n=$(NAMESPACE) -f @
       env:
         KUBECONFIG: $(KUBECONFIGFILE)
+      displayName: 'Start the test job'
```
Comment (hardbyte) [inline azure-pipelines.yml]:
```suggestion
      displayName: 'Run integration tests'
```

File: docs/devops.rst [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  - the frontend ``data61/anonlink-nginx``
+  - the backend ``data61/anonlink-app``
+  - the tutorials ``data61/anonlink-docs-tutorials`` (used to tests the tutorial Python Notebooks)
+  - the benchmark ``data61/anonlink-benchmark`` (used to run the benchmark)
+- runs the benchmark using ``docker-compose`` and publishes the artiafct in Azure
+- runs the tutorial tests using ``docker-compose`` and publishes the results in Azure
+- runs the integration tests by deploying the whole service on ``Kubernetes``, runs the tests and publishes the results in Azure.   
+
+The build pipeline is triggered for every push on every branch. It is not triggered by Pull
+Requests not to build twice the same commit.
```
Comment (hardbyte) [inline docs/devops.rst]:
```suggestion
Requests to avoid duplicate testing and building potentially untrusted code.
```

File: azure-pipelines.yml [hunk 18]
```diff
@@ -160,55 +141,78 @@ stages:
         secureFile: awsClusterConfig
     # Prepare all the variables required during the scripts.
     - template: .azurePipeline/templateDockerTagFromBranch.yml
+    # We are selecting a deployment name to be unique, not to have collision when pushing multiple times on the same branch
+    # So we are using the git commit hash. BUT, a k8s deployment MUST start with a letter, so added `es` to start its name, 
+    # then it needs to be short enough not to fail other tags built from it. In fact, most of the created objects from the templates
+    # will use the deployment name as a prefix for their tags, adding extra suffixes. If the deployment name is too long, 
+    # the tags may be longer than 63 characters which can fail the deployment with some warnings hard to see.
```
[the remark this answers is not in the document]
Comment (gusmith) [reply azure-pipelines.yml]: That was fun to find...

File: azure-pipelines.yml [hunk 19]
```diff
@@ -160,55 +141,78 @@ stages:
...
-        echo $(DOCKER_TAG)-tmppod | xargs -I@ echo "##vso[task.setvariable variable=POD_NAME]@"
-    # Start the whole service using helm. We are currently lucky, the helm version on azure is compatible with the one on our cluster, but the latest version of helm (Client 2.14.1) is not (tested from my laptop).
+        echo $(DEPLOYMENT)-tmppod | xargs -I@ echo "##vso[task.setvariable variable=POD_NAME]@"
+      displayName: 'Set some variables part 2'
+    # The file `.azurePipeline/k8s_test_pvc.yaml.tmpl` is a template for the creation of persistent volume via kubectl
+    - script: |
+        echo "Create Persistent Volume Claim"
+        cat .azurePipeline/k8s_test_pvc.yaml.tmpl | \
+        sed 's|\$PVC'"|$(PVC)|g" | \
+        sed 's|\$DEPLOYMENT_NAME'"|$(DEPLOYMENT)|g" | \
```
[the remark this answers is not in the document]
Comment (gusmith) [reply azure-pipelines.yml]: Kubectl does not accept template sadly. So the I thought about few other solutions:
 - create a new helm template but seems too much to just start a job which is only internal (I don't think we are expecting others to run integration tests)
 - updating the current helm release to include this job as a post installation hook, but I do not want to change the helm template for the deployment as it should not include or mention this integration test job
 - create a bash script to create the good files (possibly using `sed` in it, but hide them from the pipeline), or even do the whole integration test job
 - use `sed` in the pipeline and create multiple steps

I found that using helm would be too much just for that.
Then, I didn't wanted to create bash script because:
 - I don't think it could be used anywhere else than here (so not need to generalize and add documentation of all the parameters, etc)
 - described this way still allow a dev to repeat each step easily by hand if required.

In conclusion, I went for simplicity :)

File: azure-pipelines.yml (hunk 1 above)
Comment (gusmith) [reply azure-pipelines.yml]: Do we want this in this PR?
I will add `helm lint` and `helm upgrade --dry-run` for the moment.
It seems the tool you are pointing to is doing a lot more than just checking that the template is valid (e.g. version updates). So we will need to define precisely what we would like to check.

File: azure-pipelines.yml (hunk 2 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 3 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 1 above)
Comment (hardbyte) [reply azure-pipelines.yml]: Adding what you've suggested is perfect for now. thanks

File: azure-pipelines.yml (hunk 4 above)
Comment (gusmith) [reply azure-pipelines.yml]: I opened issue #393 to track this. But let's have first something working and optimize later. Or do you think this is a priority?

File: azure-pipelines.yml (hunk 5 above)
Comment (gusmith) [reply azure-pipelines.yml]: I'll clean all the test resources (the PVC, job and pod) using a label they have in common, but will keep separate the deployment cleaning in a separate step.
They currently do not have a label in common.

File: docs/devops.rst (hunk 6 above)
Comment (gusmith) [reply docs/devops.rst]: :+1:

File: docs/devops.rst (hunk 7 above)
Comment (gusmith) [reply docs/devops.rst]: :+1:

File: docs/devops.rst (hunk 8 above)
Comment (gusmith) [reply docs/devops.rst]: :+1:

File: docs/devops.rst (hunk 9 above)
Comment (gusmith) [reply docs/devops.rst]: :+1:

File: docs/devops.rst (hunk 10 above)
Comment (gusmith) [reply docs/devops.rst]: That's right: this is the kubeconfig containing the credentials to our aws cluster.

File: docs/devops.rst (hunk 11 above)
Comment (gusmith) [reply docs/devops.rst]: I put Azure pipeline because I did not want to also write the documentation for the other CI we are using, but I was expecting that at one point we should. So I'll put Continuous integration as section, and `Azure DevOps` as sub-section.

File: azure-pipelines.yml (hunk 12 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 13 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 14 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 15 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: azure-pipelines.yml (hunk 16 above)
Comment (gusmith) [reply azure-pipelines.yml]: :+1:

File: docs/devops.rst (hunk 17 above)
Comment (gusmith) [reply docs/devops.rst]: :+1:

**Comment by gusmith** [comment]
\azr help

**Comment by gusmith** [comment]
The build was failing because the cluster did not have enough resources to provide the required ones.
The k8s step failed when re-run because the variable $(DEPLOYMENT) became equal to `foo.v1`. I'm assuming that the `.v1` came from azure, but because of this, it failed as a deployment name should not include a `.`. So I added a commit to remove special characters in the deployment name.
